test/src/test/java/jenkins/model/NodesTest.java

```diff
@@ -239,4 +244,44 @@ protected boolean allowLoad(@NonNull Node node) {
         }
     }
 
+    @Test
+    public void listenersCalledOnSetNodes() throws URISyntaxException, IOException, Descriptor.FormException {
+        var agentA = new DumbSlave("nodeA", "temp", r.createComputerLauncher(null));
+        var agentB = new DumbSlave("nodeB", "temp", r.createComputerLauncher(null));
+        var agentA2 = new DumbSlave("nodeA", "temp2", r.createComputerLauncher(null));
+        Jenkins.get().setNodes(List.of(agentA, agentB));
+        assertThat(CheckSetNodes.created, containsInAnyOrder(agentA.toString(), agentB.toString()));
+        Jenkins.get().setNodes(List.of(agentA2));
+        assertThat(CheckSetNodes.updated, contains(new NodePair(agentA.toString(), agentA2.toString())));
+        assertThat(CheckSetNodes.deleted, contains(agentB.toString()));
+        Jenkins.get().setNodes(List.of());
+        assertThat(CheckSetNodes.deleted, containsInAnyOrder(agentA2.toString(), agentB.toString()));
```

Comment:
A little weak since this is not asserting the other collections in each case. Better after each `setNodes` call to assert the contents of all three collections. (Fancier test utilities have a method to return the set of recent notifications but also clear it, so the next call will be an empty list unless there is actually a _new_ notification there.)